server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierRecordingService.java

```diff
@@ -0,0 +1,149 @@
+            String action = entry.v1();
+            long timeSpentMS = entry.v2();
+
+            MeanMetric metric = recordedActions.computeIfAbsent(action, key -> new MeanMetric());
+            metric.inc(timeSpentMS);
+            seenActions.add(action);
+        }
+        recordedActions.entrySet().removeIf(entry -> seenActions.contains(entry.getKey()) == false);
+    }
+
+    static final class Recorder {
+
+        private String currentAction;
+        private long startTimeNS;
+        private boolean recording;
+        private final List<Tuple<String, Long>> recordings = new LinkedList<>();
+
+        Releasable record(String action) {
+            if (recording) {
+                throw new IllegalStateException("already recording");
+            }
+
+            this.recording = true;
+            this.currentAction = action;
+            this.startTimeNS = System.nanoTime();
+            return this::stop;
+        }
+
+        void stop() {
+            recording = false;
+            long timeSpentMS = TimeValue.nsecToMSec(System.nanoTime() - this.startTimeNS);
+            recordings.add(new Tuple<>(currentAction, timeSpentMS));
+        }
+
+        List<Tuple<String, Long>> getRecordings() {
+            return recordings;
+        }
+    }
+
+    public static class Stats implements Writeable, ToXContentFragment {
+
+        private final Map<String, Recording> recordings;
+
+        public Stats(Map<String, Recording> recordings) {
+            this.recordings = recordings;
+        }
+
+        public Stats(StreamInput in) throws IOException {
+            this(in.readMap(StreamInput::readString, Recording::new));
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject("cluster_applier_stats");
+            builder.startArray("recordings");
+            toXContentTimeSpent(builder, recordings);
+            builder.endObject();
+            return builder;
+        }
+
+        private static void toXContentTimeSpent(XContentBuilder builder, Map<String, Recording> timeSpentPerListener) throws IOException {
+            timeSpentPerListener.entrySet().stream()
+                .sorted((o1, o2) -> -Long.compare(o1.getValue().sum, o2.getValue().sum))
+                .forEach(entry -> {
+                    try {
+                        builder.startObject();
+                        builder.field("name", entry.getKey());
+                        String name = "cumulative_execution";
+                        builder.field(name + "_count", entry.getValue().count);
+                        builder.humanReadableField(name + "_time_millis", name + "_time", TimeValue.timeValueMillis(entry.getValue().sum));
+                        builder.endObject();
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
```

Comment:
Don't think I like this, we don't generally check for `UncheckedIOException`. Could we just use a normal loop and let a regular `IOException` propagate?